src/select/select_value.rs

```diff
@@ -0,0 +1,35 @@
+use crate::select::JsonPathError;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum SelectValueType {
+    Undef,
+    Null,
+    Bool,
+    Long,
+    Double,
+    String,
+    Array,
+    Dict,
+}
+
+pub trait SelectValue:
+    std::fmt::Debug + std::cmp::Eq + std::cmp::PartialEq + std::default::Default + std::clone::Clone
+{
+    fn get_type(&self) -> SelectValueType;
+    fn contains_key(&self, key: &str) -> bool;
+    fn values<'a>(&'a self) -> Option<Vec<&'a Self>>;
+    fn keys(&self) -> Option<Vec<String>>;
+    fn len(&self) -> Option<usize>;
+    fn get_key<'a>(&'a self, key: &str) -> Option<&'a Self>;
+    fn get_index<'a>(&'a self, index: usize) -> Option<&'a Self>;
```

Comment:
No need to explicitly specify the lifetime for `self` in all of these, the compiler determines it automatically.

Reply:
Are you sure? Its not compiling without it. We can look at it together on the interactive review...